pkg/envoy/xds_server.go

```diff
@@ -1320,19 +1321,43 @@ func getHTTPRule(secretManager certificatemanager.SecretManager, h *api.PortRule
 						},
 					})
 				} else {
-					// Only header presence needed
+					// Value is empty, but could be we need to set an SDS value for it instead
+					if !fromFile {
+						// Need to add a HeaderMatch so we can specify the SDS value here.
+						log.Debugf("HeaderMatches: Adding %s because SDS value is required", hdr.Name)
+						headerMatches = append(headerMatches, &cilium.HeaderMatch{
+							MismatchAction: mismatch_action,
+							Name:           hdr.Name,
+							ValueSdsSecret: namespacedNametoSyncedSDSSecretName(types.NamespacedName{
+								Namespace: hdr.Secret.Namespace,
+								Name:      hdr.Secret.Name,
+							}, policySecretsNamespace),
+						})
+					}
```

Comment:
Adding both the HeaderMatch here and the presence match below seems wrong to me. Since we don't know the value of the SDS secret, we'll have to punt the decision on the type of the match (presence only vs value match) to Envoy.

Also, we have this block above:
```
			log.WithError(err).Warning("Failed fetching K8s Secret, header match will fail")
			// Envoy treats an empty exact match value as matching ANY value; adding
			// InvertMatch: true here will cause this rule to NEVER match.
```
So we need to replicate the same "drop on a missing secret" also via SDS. (see https://github.com/cilium/proxy/pull/992)

Maybe better structure would be to check for the secret reference first, e.g., 
```
	if hdr.Secret == nil {
		value, fromFile, err := getSecretString(secretManager, hdr, ns)
		... old code as it was ...
	} else {
		log.Debugf("HeaderMatches: Adding SDS secret match for header %s", hdr.Name)
		headerMatches = append(headerMatches, &cilium.HeaderMatch{
			MismatchAction: mismatch_action,
			Name:           hdr.Name,
			ValueSdsSecret: namespacedNametoSyncedSDSSecretName(types.NamespacedName{
				Namespace: hdr.Secret.Namespace,
				Name:      hdr.Secret.Name,
			}, policySecretsNamespace),
		})
	}
	...
```

Reply:
The problem with the approach as written is that `hdr.Secret` is only `nil` if no Secret is specified at all; that is, if the value is specified directly. If the value is being read from the file system, then `hdr.Secret` contains a `name` and `namespace` that are used to build the path to read from instead.

But I agree that the logic is a bit convoluted, I'll see if I can tidy it up a bit.

Reply:
Done in 2652aedc2473215bbeaa4b502358d315673ed5d8.